### ISSUE TITLE
Deleting channels should remove sock js listeners

### ISSUE THREAD

**Issue by jkimbo** [FelixOnline/Sprinkler#4]
When deleting a channel the sockjs listener for that channel should also be removed.

**Comment by lsproc** [comment]
@jkimbo what is your opinion of this fix? I am not sure whether this is a good fix from a scalability stand point although I do wonder how often a channel will be deleted...

**Comment by jkimbo** [comment]
@lsproc I'm not quite sure I follow it. Couldn't you keep references to each sock.js connection in memory and then when deleting an endpoint just kill the connection too?

**Comment by lsproc** [comment]
I don't think so. Let me try and explain a bit more clearly what's going on:

The issue is when the SockJS handlers are installed, they are basically pushed onto the list of handlers currently assigned to Express - so one cannot just pop them off as there may be new handlers installed since.

You cannot go in and remove the handler progmatically as it seems to be a list of anonymous functions rather than an object that we could just go in and remove the key for.

Nor can we uninstall the handlers and put the original ones back as that will delete the handlers for all the other SockJS instances as well.

Now, in terms of deleting the SockJS connection, I don't believe there is any option to close a server instance.  However, I have since this commit started storing each Socket instance in memory and deleting them when the endpoint is closed - so hopefully Node's garbage collection being effective at making sure that the memory is cleaned up.

We will still need to do something about the handlers however as I found that if you have handlers pointing to nowhere then Express seems to hang. So, the commit I introduced of installing handlers to serve a 404 nicely deals with that at the expense of having a messy list of handlers with some unused ones sitting around. I'm not sure how that will scale - it could be fine but I just don't know...

**Comment by jkimbo** [comment]
@lsproc hmm ok seems reasonable to me. I guess we're probably abusing sock.js here a bit since I don't think dynamic handlers is something that was considered.

**Comment by lsproc** [comment]
I get the impression that this isn't just a sockjs thing - it looks like its something that would need to be done in the Node events module too.

**Comment by entrepreneurj** [comment]
Just thought I'd have a look into this myself, since I had some similar issues with a websockets server I had a while ago. If I understand what you're saying, I think you're fine with your handlers, but I think in the protocol they officially send further close events to clients looking for a closed endpoint/channel. http://sockjs.github.io/sockjs-protocol/sockjs-protocol-0.3.3.html
Happy socking!

**Comment by lsproc** [comment]
Good to hear from you, Joe!

That is quite interesting to read - thanks for that. However, that seems to imply that the SockJS server is responsible for sending these events (in this case, sockjs-node rather than Sprinkler as Sprinkler is simply a wrapper around it and an API). I cannot see anything in sockjs-node to close and endpoint. Although this could just mean that sockjs-node is not compliant!

**Comment by lsproc** [comment]
Ahhh no I just missed the relevant part of the docs.

I can close each connection to the Socket when the socket is closed and send a message. Do we then want to dispose of the socket, or keep it and have it immediately firing off Close events? I'm not sure what the advantages/disadvantages of each are so if you have more experience in this I'd appreciate your advice.

**Comment by entrepreneurj** [comment]
Yeah it's a tough one isn't it! To be fair I'm not entirely sure, I'm a bit shaky here. I'd say a closed socket is probably the way to go, but I don't know the memory implications (if any) of that. It might just be easier to dispose of the closed socket and _404_. There may not be much difference, I'd hope that clients trying to access closed wsockets would be a rare occurrence :) It's probably (hopefully) only a significant difference in specific use cases. 
Sorry I couldn't be of any further use! :)
